src/types.rs

```diff
@@ -1088,6 +1090,121 @@ impl SimpleAsn1Writable for X509GeneralizedTime {
+        if let Some(val) = nanoseconds {
+            if val < 1 || val >= 1e9 as u32 {
+                return Err(ParseError::new(ParseErrorKind::InvalidValue));
+            }
+        }
+
+        Ok(GeneralizedTime {
+            datetime: dt,
+            nanoseconds,
+        })
+    }
+
+    pub fn as_datetime(&self) -> &DateTime {
+        &self.datetime
+    }
+
+    pub fn nanoseconds(&self) -> Option<u32> {
+        self.nanoseconds
+    }
+}
+
+fn read_fractional_time(data: &mut &[u8]) -> ParseResult<Option<u32>> {
+    // We cannot use read_byte here because it will advance the pointer
+    // However, we know that the is suffixed by 'Z' so reading into an empty
+    // data should lead to an error.
+    if data.first() == Some(&b'.') {
+        *data = &data[1..];
+
+        let mut fraction = 0u32;
+        let mut digits = 0;
+        // Read up to 9 digits
+        for b in data.iter().take(9) {
+            if !b.is_ascii_digit() {
+                if digits == 0 {
+                    // We must have at least one digit
+                    return Err(ParseError::new(ParseErrorKind::InvalidValue));
+                }
+                break;
+            }
+            fraction = fraction * 10 + (b - b'0') as u32;
+            digits += 1;
+        }
+        *data = &data[digits..];
+
+        // No trailing zero
+        if fraction % 10 == 0 {
+            return Err(ParseError::new(ParseErrorKind::InvalidValue));
+        }
+
+        // Now let scale up in nanoseconds
+        let nanoseconds: u32 = fraction * 10u32.pow(9 - digits as u32);
+        Ok(Some(nanoseconds))
+    } else {
+        Ok(None)
+    }
+}
+
+impl SimpleAsn1Readable<'_> for GeneralizedTime {
+    const TAG: Tag = Tag::primitive(0x18);
+    fn parse_data(mut data: &[u8]) -> ParseResult<GeneralizedTime> {
+        let year = read_4_digits(&mut data)?;
+        let month = read_2_digits(&mut data)?;
+        let day = read_2_digits(&mut data)?;
+        let hour = read_2_digits(&mut data)?;
+        let minute = read_2_digits(&mut data)?;
+        let second = read_2_digits(&mut data)?;
+
+        let fraction = read_fractional_time(&mut data)?;
+        read_tz_and_finish(&mut data)?;
+
+        GeneralizedTime::new(
+            DateTime::new(year, month, day, hour, minute, second)?,
+            fraction,
+        )
+    }
+}
+
+impl SimpleAsn1Writable for GeneralizedTime {
+    const TAG: Tag = Tag::primitive(0x18);
+    fn write_data(&self, dest: &mut WriteBuf) -> WriteResult {
+        let dt = self.as_datetime();
+        push_four_digits(dest, dt.year())?;
+        push_two_digits(dest, dt.month())?;
+        push_two_digits(dest, dt.day())?;
+
+        push_two_digits(dest, dt.hour())?;
+        push_two_digits(dest, dt.minute())?;
+        push_two_digits(dest, dt.second())?;
+
+        if let Some(nanoseconds) = self.nanoseconds() {
+            dest.push_byte(b'.')?;
+
+            for digit in format!("{:09}", nanoseconds)
```

Comment:
Rather than use `format!()`, which will allocate a new string, let's use `write!()` with a buffer on the stack:

Reply:
I might be missing something, but I think this will require `use std::io::Write`, i.e. break the current no-std assumption.

For example:

```rust
            let mut buf = [0_u8; 9];
            write!(&mut buf[..], "{nanoseconds:09}").unwrap();
            dest.push_slice(&buf)?;
```

(that doesn't handle the trimming yet, but it gets the demonstrates that `Write` is needed to write into a `&mut [u8]`.)

Reply:
Grumble, is there really no core/alloc API to let us write an integer into a stack buf?

Reply:
Not that I know of 🙃 -- I could do this the ugly way if you'd like, by manually mashing values into a stack buf. I can play around with that a bit.

Reply:
Annoying. We currently basically always avoid allocations unless the caller is using a particular owned type, so annoyingly I do think this is something we want to do.

Reply:
I've added `atoi` as a dependency to do this, after confirming that it's `no_std` and performs admirably. LMK what you think!